pkg/app/pipedv1/cmd/piped/grpcapi/tool_registry.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcapi
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"text/template"
+
+	"golang.org/x/sync/singleflight"
+)
+
+type toolRegistry struct {
+	toolsDir string
+	tmpDir   string
+	group    singleflight.Group
+}
```

Comment:
Since the tool is not plugin-specific, I think we should have some kind of cache to avoid redundant installation. If the name and version are the same, then we should return the installed path immediately instead of running the script, wdyt?

Reply:
I agree. I'll add the cache like below.
https://github.com/pipe-cd/pipecd/blob/23eddac42ab78523e522ac534437b8718f200ba1/pkg/app/piped/toolregistry/registry.go#L116-L121

Reply:
@khanhtc1202 
fixed on this commit, thanks
https://github.com/pipe-cd/pipecd/pull/5343/commits/75d7daaf84f259431f11e67ec79472b3b7d9025d